### PR TITLE
cakelisp: init at 0.1.0

### PR DESCRIPTION
**Pull request by sbond75**
###### Motivation for this change

Add cakelisp and add myself as a maintainer.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 107439` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cakelisp</li>
  </ul>
</details>

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -7921,6 +7921,12 @@
     githubId = 720864;
     name = "Sébastien Bourdeauducq";
   };
+  sbond75 = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:7924]: Please put this into a separate commit with the message ``maintainers: add sbond75``.

File: pkgs/development/compilers/cakelisp/default.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  '';
+
+  buildPhase = ''jam "-j$NIX_BUILD_CORES"'';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp bin/cakelisp $out/bin/
+  '';
+
+  meta = with lib; {
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/cakelisp/default.nix:45]:
```suggestion
  meta = with stdenv.lib; {
```

File: pkgs/development/compilers/cakelisp/default.nix [hunk 3]
```diff
@@ -0,0 +1,52 @@
+{ stdenv, lib, fetchFromGitHub, clang, jam }:
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/cakelisp/default.nix:1]:
```suggestion
{ stdenv, fetchFromGitHub, clang, jam }:
```

File: pkgs/development/compilers/cakelisp/default.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    rev = "226bb4385a6fb6d61d37a5be8100004a140e5f65";
+    sha256 = "05fyj9a8m44kc3lzyqz4a8dnwa2rpcq981ryszsacywsng46nb01";
+  };
+
+  nativeBuildInputs = [ jam ];
+
+  buildInputs = [ clang ];
+
+  postPatch = ''
+    substituteInPlace src/Logging.hpp --replace 'log;' 'gLog;'
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/cakelisp/default.nix:19]: Is there a patch upstream which fixes this?

File: pkgs/development/compilers/cakelisp/default.nix [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        --replace '"/usr/bin/clang++"' '"${clang}/bin/clang++"'
+    substituteInPlace src/ModuleManager.cpp \
+        --replace '"/usr/bin/clang++"' '"${clang}/bin/clang++"'
+  '';
+
+  buildPhase = ''jam "-j$NIX_BUILD_CORES"'';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp bin/cakelisp $out/bin/
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/cakelisp/default.nix:42]: Using install here would be great.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 107439` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cakelisp</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 107439` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cakelisp</li>
  </ul>
</details>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

`gpl3` is a deprecated license, check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

File: pkgs/development/compilers/cakelisp/default.nix (hunk 4 above)
Comment (siraben) [reply pkgs/development/compilers/cakelisp/default.nix:19]: Not at the moment, it appears that this substitution is needed to fix the symbol conflict on darwin (where it conflicts with `math.h`'s `log`).

Comment (SuperSandro2000) [reply pkgs/development/compilers/cakelisp/default.nix:19]: Could you create a upstream PR and fetch that?

File: pkgs/development/compilers/cakelisp/default.nix (hunk 3 above)
Comment (sbond75) [reply pkgs/development/compilers/cakelisp/default.nix:1]: This shouldn't be needed because I used lib.optionalString

File: pkgs/development/compilers/cakelisp/default.nix (hunk 4 above)
Comment (sbond75) [reply pkgs/development/compilers/cakelisp/default.nix:19]: This could be done in the future, repo is changing a lot currently

File: pkgs/development/compilers/cakelisp/default.nix (hunk 2 above)
Comment (sbond75) [reply pkgs/development/compilers/cakelisp/default.nix:44]: This shouldn't be needed because I imported lib

Comment (SuperSandro2000) [reply pkgs/development/compilers/cakelisp/default.nix:44]: This is some cleanup to remove lib from inputs.

File: pkgs/development/compilers/cakelisp/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/development/compilers/cakelisp/default.nix:19]: Either do it now or never. I don't think anyone will remember this in 2 weeks.

**Comment by siraben** [comment]
Tested on darwin, the chain of `substituteInPlace` seems unnecessary now.

```diff
diff --git a/pkgs/development/compilers/cakelisp/default.nix b/pkgs/development/compilers/cakelisp/default.nix
index a42a05cbc68..6f40ac126ba 100644
--- a/pkgs/development/compilers/cakelisp/default.nix
+++ b/pkgs/development/compilers/cakelisp/default.nix
@@ -1,38 +1,27 @@
-{ stdenv, lib, fetchFromGitHub, clang, jam }:
+{ stdenv, lib, fetchFromGitHub, gcc, jam }:
 
 stdenv.mkDerivation {
   pname = "cakelisp";
-  version = "unstable-2020-12-21";
+  version = "unstable-2020-12-23";
 
   src = fetchFromGitHub {
     owner = "makuto";
     repo = "cakelisp";
-    rev = "226bb4385a6fb6d61d37a5be8100004a140e5f65";
-    sha256 = "05fyj9a8m44kc3lzyqz4a8dnwa2rpcq981ryszsacywsng46nb01";
+    rev = "1754288817ea325b466a6b80d2c29beff6f5580a";
+    sha256 = "1l7vg4y3ysxdjk3hn8mbhcnw3k3lnpa268j39k8m0m7b45p6lgwa";
   };
 
   nativeBuildInputs = [ jam ];
 
-  buildInputs = [ clang ];
+  buildInputs = [ gcc ];
 
   postPatch = ''
-    substituteInPlace src/Logging.hpp --replace 'log;' 'gLog;'
-    substituteInPlace src/Logging.cpp --replace 'log =' 'gLog ='
-    substituteInPlace src/Utilities.cpp --replace 'log.' 'gLog.'
-    substituteInPlace src/Tokenizer.cpp --replace 'log.' 'gLog.'
-    substituteInPlace src/FileUtilities.cpp --replace 'log.' 'gLog.'
-    substituteInPlace src/Main.cpp --replace 'log.' 'gLog.'
-    substituteInPlace src/Evaluator.cpp --replace 'log.' 'gLog.'
-    substituteInPlace src/Writer.cpp --replace 'log.' 'gLog.'
-    substituteInPlace src/RunProcess.cpp --replace 'log.' 'gLog.'
-    substituteInPlace src/ModuleManager.cpp --replace 'log.' 'gLog.'
-
     ${lib.optionalString stdenv.isDarwin "substituteInPlace Jamrules --replace 'OS_DEPENDENT_LINKFLAGS = --export-dynamic ;' 'OS_DEPENDENT_LINKFLAGS = -export_dynamic ;'"}
 
     substituteInPlace runtime/HotReloading.cake \
-        --replace '"/usr/bin/clang++"' '"${clang}/bin/clang++"'
+        --replace '"/usr/bin/g++"' '"${gcc}/bin/g++"'
     substituteInPlace src/ModuleManager.cpp \
-        --replace '"/usr/bin/clang++"' '"${clang}/bin/clang++"'
+        --replace '"/usr/bin/g++"' '"${gcc}/bin/g++"'
   '';
 
   buildPhase = ''jam "-j$NIX_BUILD_CORES"'';
```

File: pkgs/development/compilers/cakelisp/default.nix [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    owner = "makuto";
+    repo = "cakelisp";
+    rev = "1754288817ea325b466a6b80d2c29beff6f5580a";
+    sha256 = "1l7vg4y3ysxdjk3hn8mbhcnw3k3lnpa268j39k8m0m7b45p6lgwa";
+  };
+
+  buildInputs = [ gcc ];
+
+  postPatch = ''
+
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/cakelisp/default.nix:17]:
```suggestion
```

File: pkgs/development/compilers/cakelisp/default.nix [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    ${stdenv.lib.optionalString stdenv.isDarwin "substituteInPlace Build.sh --replace '--export-dynamic' '-export_dynamic'"}
+    ${stdenv.lib.optionalString stdenv.isDarwin "substituteInPlace runtime/HotReloading.cake --replace '--export-dynamic' '-export_dynamic'"}
+    ${stdenv.lib.optionalString stdenv.isDarwin "substituteInPlace Bootstrap.cake --replace '--export-dynamic' '-export_dynamic'"}
+
+    substituteInPlace runtime/HotReloading.cake \
+        --replace '"/usr/bin/g++"' '"${gcc}/bin/g++"'
+    substituteInPlace src/ModuleManager.cpp \
+        --replace '"/usr/bin/g++"' '"${gcc}/bin/g++"'
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/cakelisp/default.nix:26]:
```suggestion
    substituteInPlace runtime/HotReloading.cake \
        --replace '"/usr/bin/g++"' '"${gcc}/bin/g++"'
    substituteInPlace src/ModuleManager.cpp \
        --replace '"/usr/bin/g++"' '"${gcc}/bin/g++"'
  '' + stdenv.lib.optionalString stdenv.isDarwin ''
    substituteInPlace Build.sh --replace '--export-dynamic' '-export_dynamic'
    substituteInPlace runtime/HotReloading.cake --replace '--export-dynamic' '-export_dynamic'
    substituteInPlace Bootstrap.cake --replace '--export-dynamic' '-export_dynamic'
    '';
```

Didn't test but repeating optionalString three times is a bit repetitive.

File: pkgs/development/compilers/cakelisp/default.nix [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  buildPhase = "./Build.sh";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/cakelisp/default.nix:28]:
```suggestion
  buildPhase = ''
    ./Build.sh
  '';
```

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 107439` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cakelisp</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 107439` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cakelisp</li>
  </ul>
</details>

**Comment by siraben** [comment]
@sbond75 please update the version to v0.1.0 (and refresh the sha256) https://github.com/makuto/cakelisp/releases

**Review by siraben — APPROVED** [review]

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 107439` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cakelisp</li>
  </ul>
</details>
